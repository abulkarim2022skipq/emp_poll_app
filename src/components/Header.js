import { connect } from "react-redux";
import { Link, useNavigate } from "react-router-dom";

const Header = (props) => {
  const { nav, user } = props;

  const navigate = useNavigate();

  const handleLogout = (e) => {
    e.preventDefault();

    navigate("/login");
  };

  return (
    <header>
      <nav className="nav">
        <div className="nav-div">
          <div className="nav-left">
            <ul>
              <li>
                <Link to="/" className={nav === "home" ? "active-nav" : ""}>
                  Home
                </Link>
              </li>
              <li>
                <Link
                  to="/leaderboard"
                  className={nav === "leaderboard" ? "active-nav" : ""}
                >
                  Leaderboard
                </Link>
              </li>
              <li>
                <Link to="/add" className={nav === "new" ? "active-nav" : ""}>
                  New Question
                </Link>
              </li>
            </ul>
          </div>
          <div className="nav-right">
            <ul>
              <li className="avatar-smaller">
                <Link to={"/"}>
                  {user !== undefined ? (
                    <img src={user.avatarURL} width={"50px"} />
                  ) : (
                    ""
                  )}
                </Link>
              </li>
              <li>
                <Link to={"/"}>{user !== undefined ? user.id : ""}</Link>
              </li>
              <li>
                <a onClick={handleLogout} to={"/login"}>
                  Logout
                </a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </header>
  );
};

const mapStateToProps = ({ nav, users, authedUser }) => {
  return { nav, user: users[authedUser] };
};

export default connect(mapStateToProps)(Header);
