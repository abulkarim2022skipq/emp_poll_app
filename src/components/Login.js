import { useEffect, useState } from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import { setAuthedUser } from "../actions/authedUser";

const Login = (props) => {
  const { users, dispatch } = props;

  const navigate = useNavigate();

  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [wrongPasswordAlert, setWrongPasswordAlert] = useState(false);
  const [wrongPassword, setWrongPassword] = useState(false);

  const handleSubmit = (e) => {
    e.preventDefault();

    try {
      if (users[username] !== undefined) {
        if (users[username].password === password) {
          dispatch(setAuthedUser(username));
          navigate("/");
        }
      }
      setWrongPassword(true);
    } catch (error) {
      setWrongPassword(true);
      console.log(error);
    }
  };

  useEffect(() => {
    if (wrongPassword) {
      setWrongPasswordAlert(true);
      setUsername("");
      setPassword("");
      const timeout = setTimeout(() => {
        setWrongPasswordAlert(false);
        setWrongPassword(false);
      }, 3000);
      return () => {
        clearInterval(timeout);
      };
    }
  }, [wrongPassword]);

  return (
    <div className="login">
      <div className="center">
        <h3>Login</h3>
      </div>
      <form onSubmit={handleSubmit}>
        <div className="login-options">
          <label>Username:</label>
          <input
            data-testid="username"
            type="text"
            placeholder="Username"
            value={username}
            onChange={(e) => {
              setUsername(e.target.value);
            }}
          />
        </div>
        <div className="login-options">
          <label>Password:</label>
          <input
            data-testid="password"
            type="password"
            placeholder="Password"
            value={password}
            onChange={(e) => {
              setPassword(e.target.value);
            }}
          />
        </div>
        <div className="login-submit-button">
          <div>
            <button
              className="question-show-button"
              type="submit"
              disabled={
                username.length < 1 || password.length < 1 ? true : false
              }
            >
              Login
            </button>
          </div>
        </div>
      </form>
      {wrongPasswordAlert && (
        <div className="wrong-password-alert-div">
          <div className="wrong-password-alert">
            Username or Password is incorrect
          </div>
        </div>
      )}
    </div>
  );
};
const mapStateToProps = ({ users }) => {
  return {
    users,
  };
};
export default connect(mapStateToProps)(Login);
