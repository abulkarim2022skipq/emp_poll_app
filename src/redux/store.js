import { createStore } from "redux";
import reducers from "../reducers";
import middleware from "../middleware";

const store = createStore(
  reducers,
  middleware
  // window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);

export default store;
