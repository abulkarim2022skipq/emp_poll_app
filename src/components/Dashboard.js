import { useEffect } from "react";
import { connect } from "react-redux";
import { setPageLocation } from "../actions/nav";
import QuestionContainer from "./QuestionContainer";
import Header from "./Header";

const Dashboard = (props) => {
  const { newQuestions, doneQuestions, dispatch } = props;

  useEffect(() => {
    dispatch(setPageLocation("home"));
  }, [dispatch]);

  return (
    <div>
      <Header />
      <div>
        <QuestionContainer title="New Questions" questions={newQuestions} />
        <QuestionContainer title="Done" questions={doneQuestions} />
      </div>
    </div>
  );
};
const mapStateToProps = ({ authedUser, users, questions }) => {
  const newQuestions = Object.keys(questions).filter(
    (que) =>
      que !== Object.keys(users[authedUser].answers).find((e) => e === que)
  );

  const doneQuestions = Object.keys(questions).filter(
    (que) =>
      que === Object.keys(users[authedUser].answers).find((e) => e === que)
  );

  return {
    newQuestions,
    doneQuestions,
  };
};
export default connect(mapStateToProps)(Dashboard);
