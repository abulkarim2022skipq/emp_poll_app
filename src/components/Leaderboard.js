import { useEffect } from "react";
import { connect } from "react-redux";
import { setPageLocation } from "../actions/nav";
import Header from "./Header";

const Leaderboard = (props) => {
  const { dispatch, users } = props;

  useEffect(() => {
    dispatch(setPageLocation("leaderboard"));
  }, []);

  if (users === undefined || users.length < 1) {
    return <div>No data found</div>;
  }

  return (
    <div>
      <Header />

      <div className="leaderboard">
        <table className="leaderboard-table">
          <thead>
            <tr>
              <td>User</td>
              <td>Answered</td>
              <td>Created</td>
            </tr>
          </thead>
          <tbody>
            {users.length < 1 === true ? (
              <div>Not</div>
            ) : (
              users.map((user) => (
                <tr key={user.id}>
                  <td>
                    <div className="leaderboard-user">
                      <div className="avatar-small">
                        <img
                          src={user.avatarURL}
                          alt="avatar"
                          width="50px"
                          height="50px"
                        />
                      </div>
                      <div className="leaderboard-user-details">
                        <div className="leaderboard-username">{user.name}</div>
                        <div className="leaderboard-userid">{user.id}</div>
                      </div>
                    </div>
                  </td>
                  <td>{Object.keys(user.answers).length}</td>
                  <td>{user.questions.length}</td>
                </tr>
              ))
            )}
          </tbody>
        </table>
      </div>
    </div>
  );
};

const mapStateToProps = ({ users }) => {
  var usersToReturn = Object.keys(users).map((user) => {
    const { answers, name, questions, id, avatarURL } = users[user];
    return { answers, name, questions, id, avatarURL };
  });

  usersToReturn = usersToReturn.sort((a, b) => {
    const sumA = a.questions.length + Object.keys(a.answers).length;
    const sumB = b.questions.length + Object.keys(b.answers).length;
    return sumB - sumA;
  });

  return { users: usersToReturn };
};

export default connect(mapStateToProps)(Leaderboard);
