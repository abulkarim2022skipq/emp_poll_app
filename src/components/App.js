import { Fragment, useEffect } from "react";
import { connect } from "react-redux";
import { handleInitialData } from "../actions/shared";
import { Route, Routes, useNavigate } from "react-router-dom";
import Dashboard from "./Dashboard";
import NotFound from "./NotFound";
import Leaderboard from "./Leaderboard";
import NewQuestion from "./NewQuestion";
import QuestionPage from "./QuestionPage";
import Login from "./Login";

function App(props) {
  const { dispatch, loading } = props;

  const navigate = useNavigate();

  useEffect(() => {
    dispatch(handleInitialData());
    if (loading === false) {
      navigate("/login");
    }
  }, []);

  return (
    <Fragment>
      <div className="container">
        {loading === true ? null : (
          <div>
            <Routes>
              <Route path="/" exact element={<Dashboard />} />
              <Route path="/question/:id" element={<QuestionPage />} />
              <Route path="/leaderboard" element={<Leaderboard />} />
              <Route path="/add" element={<NewQuestion />} />
              <Route path="/login" element={<Login />} />
              <Route path="/*" element={<NotFound />} />
            </Routes>
          </div>
        )}
      </div>
    </Fragment>
  );
}

const mapStateToProps = ({ authedUser }) => {
  return {
    loading: authedUser === null,
  };
};

export default connect(mapStateToProps)(App);
