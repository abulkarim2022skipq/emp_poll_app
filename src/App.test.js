import {
  _saveQuestion,
  _saveQuestionAnswer,
  _getQuestions,
  _getUsers,
} from "./utils/_DATA";
import React from "react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import { fireEvent, render } from "@testing-library/react";
import { toBeInTheDocument } from "@testing-library/jest-dom";
import store from "./redux/store";
import App from "./components/App";
import Login from "./components/Login";
import Header from "./components/Header";
import Dashboard from "./components/Dashboard";
import Leaderboard from "./components/Leaderboard";

describe("testing async function SaveQuestion", () => {
  it("will return an object if correct values are passed", async () => {
    let question = {
      optionOneText: "learn French",
      optionTwoText: "learn Italian",
      author: "sarahedo",
    };
    let result = await _saveQuestion(question);
    expect(result.optionOne.text).toEqual(question.optionOneText);
    expect(result.optionTwo.text).toEqual(question.optionTwoText);
    expect(result.author).toEqual(question.author);
  });

  it("will reject the promise and throw error if incorrect values are passed", async () => {
    let question = {
      optionOneText: "learn a new language",
      optionTwoText: "learn a new framework",
    };
    await expect(_saveQuestion(question)).rejects.toEqual(
      "Please provide optionOneText, optionTwoText, and author"
    );
  });
});

describe("testing async function SaveQuestionAnswer", () => {
  it("will resolve with true if correct values are passed", async () => {
    let authedUser = "sarahedo";
    let questionid = "6ni6ok3ym7mf1p33lnez";
    let answer = "optionOne";
    let result = await _saveQuestionAnswer({
      authedUser: authedUser,
      qid: questionid,
      answer: answer,
    });
    expect(result).toEqual(true);
  });

  it("will reject the promise and throw error if incorrect values are passed", async () => {
    let authedUser = "sarahedo";

    await expect(
      _saveQuestionAnswer({
        authedUser: authedUser,
      })
    ).rejects.toEqual("Please provide authedUser, qid, and answer");
  });
});

describe("testing users and questions from backend(_DATA.js)", () => {
  it("will return all users from database", async () => {
    await expect(_getUsers()).not.toBeNull();
  });

  it("will return all questions from database", async () => {
    await expect(_getQuestions()).not.toBeNull();
  });
});

describe("testing snapshot of different components", () => {
  it("App component", () => {
    let component = render(
      <MemoryRouter>
        <Provider store={store}>
          <App />
        </Provider>
      </MemoryRouter>
    );
    expect(component).toMatchSnapshot();
  });

  it("Leaderboard Component", () => {
    let component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Leaderboard />
        </Provider>
      </MemoryRouter>
    );
    expect(component).toMatchSnapshot();
  });

  it("Header ", () => {
    let component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Header />
        </Provider>
      </MemoryRouter>
    );
    expect(component).toMatchSnapshot();
  });
});

describe("testing DOM", () => {
  it("Login", () => {
    let component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
          <Dashboard />
        </Provider>
      </MemoryRouter>
    );
    let userid = component.getByTestId("username");
    fireEvent.change(userid, { target: { value: "tylermcginnis" } });
    let password = component.getByTestId("password");
    fireEvent.change(password, { target: { value: "abc321" } });
    let submitButton = component.getByRole("button");
    fireEvent.click(submitButton);
    let dashboard = component.getByTestId("dashboard");
    expect(dashboard).toBeInTheDocument();
  });
});
