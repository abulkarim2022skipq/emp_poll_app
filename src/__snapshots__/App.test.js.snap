// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing snapshot of different components App component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div>
          <div
            class="login"
          >
            <div
              class="center"
            >
              <h3>
                Login
              </h3>
            </div>
            <form>
              <div
                class="login-options"
              >
                <label>
                  Username:
                </label>
                <input
                  data-testid="username"
                  placeholder="Username"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="login-options"
              >
                <label>
                  Password:
                </label>
                <input
                  data-testid="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="login-submit-button"
              >
                <div>
                  <button
                    class="question-show-button"
                    disabled=""
                    type="submit"
                  >
                    Login
                  </button>
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div>
        <div
          class="login"
        >
          <div
            class="center"
          >
            <h3>
              Login
            </h3>
          </div>
          <form>
            <div
              class="login-options"
            >
              <label>
                Username:
              </label>
              <input
                data-testid="username"
                placeholder="Username"
                type="text"
                value=""
              />
            </div>
            <div
              class="login-options"
            >
              <label>
                Password:
              </label>
              <input
                data-testid="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="login-submit-button"
            >
              <div>
                <button
                  class="question-show-button"
                  disabled=""
                  type="submit"
                >
                  Login
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`testing snapshot of different components Header  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header>
        <nav
          class="nav"
        >
          <div
            class="nav-div"
          >
            <div
              class="nav-left"
            >
              <ul>
                <li>
                  <a
                    class=""
                    href="/"
                  >
                    Home
                  </a>
                </li>
                <li>
                  <a
                    class="active-nav"
                    href="/leaderboard"
                  >
                    Leaderboard
                  </a>
                </li>
                <li>
                  <a
                    class=""
                    href="/add"
                  >
                    New Question
                  </a>
                </li>
              </ul>
            </div>
            <div
              class="nav-right"
            >
              <ul>
                <li
                  class="avatar-smaller"
                >
                  <a
                    href="/"
                  />
                </li>
                <li>
                  <a
                    href="/"
                  />
                </li>
                <li>
                  <a
                    to="/login"
                  >
                    Logout
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </nav>
      </header>
    </div>
  </body>,
  "container": <div>
    <header>
      <nav
        class="nav"
      >
        <div
          class="nav-div"
        >
          <div
            class="nav-left"
          >
            <ul>
              <li>
                <a
                  class=""
                  href="/"
                >
                  Home
                </a>
              </li>
              <li>
                <a
                  class="active-nav"
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
              </li>
              <li>
                <a
                  class=""
                  href="/add"
                >
                  New Question
                </a>
              </li>
            </ul>
          </div>
          <div
            class="nav-right"
          >
            <ul>
              <li
                class="avatar-smaller"
              >
                <a
                  href="/"
                />
              </li>
              <li>
                <a
                  href="/"
                />
              </li>
              <li>
                <a
                  to="/login"
                >
                  Logout
                </a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`testing snapshot of different components Leaderboard Component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        No data found
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      No data found
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
